keps/sig-network/2104-reworking-kube-proxy-architecture/README.md

```diff
@@ -0,0 +1,782 @@
+includes a default Run function, setting up default flags, parsing them and running the client,
+allowing very simple clients like this:
+
+```golang
+package main
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/mcluseau/kube-proxy2/pkg/api/localnetv1"
+	"github.com/mcluseau/kube-proxy2/pkg/client"
+)
+
+func main() {
+	client.Run(printState)
+}
+
+func printState(items []*localnetv1.ServiceEndpoints) {
+	fmt.Fprintln(os.Stdout, "#", time.Now())
+	for _, item := range items {
+		fmt.Fprintln(os.Stdout, item)
+	}
+}
+```
+
+The currently proposed interface for the lower-level client is as follows:
+
+```godoc
+package client // import "github.com/mcluseau/kube-proxy2/pkg/client"
+
+type EndpointsClient struct {
+	// Target is the gRPC dial target
+	Target string
+
+	// InstanceID and Rev are the latest known state (used to resume a watch)
+	InstanceID uint64
+	Rev        uint64
+
+	// ErrorDelay is the delay before retrying after an error.
+	ErrorDelay time.Duration
+
+	// Has unexported fields.
+}
+    EndpointsClient is a simple client to kube-proxy's Endpoints API.
+
+func New(flags FlagSet) (epc *EndpointsClient)
+    New returns a new EndpointsClient with values bound to the given flag-set
+    for command-line tools. Other needs can use `&EndpointsClient{...}`
+    directly.
+
+func (epc *EndpointsClient) Cancel()
+    Cancel will cancel this client, quickly closing any call to Next.
+
+func (epc *EndpointsClient) CancelOn(signals ...os.Signal)
+    CancelOn make the given signals to cancel this client.
+
+func (epc *EndpointsClient) CancelOnSignals()
+    CancelOnSignals make the default termination signals to cancel this client.
+
+func (epc *EndpointsClient) DefaultFlags(flags FlagSet)
+    DefaultFlags registers this client's values to the standard flags.
+
+func (epc *EndpointsClient) Next() (items []*localnetv1.ServiceEndpoints, canceled bool)
+    Next returns the next set of ServiceEndpoints, waiting for a new revision as
+    needed. It's designed to never fail and will always return latest items,
+    unless canceled.
+```
+
+A good example of how to use this low level API is the `client.Run` itself:
+
+```golang
+// Run the client with the standard options
+func Run(handlers ...HandlerFunc) {
+	once := flag.Bool("once", false, "only one fetch loop")
+
+	epc := New(flag.CommandLine)
+
+	flag.Parse()
+
+	epc.CancelOnSignals()
+
+	for {
+		items, canceled := epc.Next()
+
+		if canceled {
+			return
+		}
+
+		for _, handler := range handlers {
+			handler(items)
+		}
+
+		if *once {
+			klog.Infof("to resume this watch, use --instance-id %d --rev %d", epc.InstanceID, epc.Rev)
+			return
+		}
+	}
+}
+```
+
+- use the docker syntax to express binding, allowing sockets with
+  `unix:///run/kubernetes/proxy.sock`
+- may economize some syscalls for internal implementations by using
+  `google.golang.org/grpc/test/bufconn`, but that sounds like premature optimization
+
+### Test Plan
```

Comment:
This _really_ needs to be filled in.

A huge percentage of kubernetes users depend on the iptables proxy. How are we going to convince ourselves that the very-heavily-rewritten version in kpng is correct and non-buggy?

Reply:
right now in CI we run roughly... 

```
-> % /home/ubuntu/SOURCE/kubernetes/e2e.test --provider=local --kubeconfig=/home/ubuntu/.kube/config --dump-logs-on-failure=false --ginkgo.focus="sig-network" --ginkgo.skip="Driver|Slow|Driver" --ginkgo.dryRun=true
```

and these are mostly passing.  We dont have explicit scalability tests that we run yet though, if theres a large scale 1 time kube proxy scale test we should run, let us know ? maybe something in kube-loader ?

Reply:
Per conv w/ dan, we'll iterate on this....